Release 5.0.0

abhishek-gupta-nuxeo opened the pull request:
This pull request will publish release 5.0.0 for 2025 version.

abhishek-gupta-nuxeo committed:
Release 5.0.0

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/nuxeo/nuxeo-ai?pullRequest=665) <br/>All committers have signed the CLA.

debjit-hyland commented:
We need to upgrade the nuxeo-parent to 2025 as well.

abhishek-gupta-nuxeo committed:
aws sdk upgrade

abhishek-gupta-nuxeo committed:
search dependency upgrade

abhishek-gupta-nuxeo committed:
junits tests fixed

abhishek-gupta-nuxeo committed:
Merge branch '2025' into task-NXCON-61-release-nuxeo-ai-2025

abhishek-gupta-nuxeo committed:
junit fix

kevinleturc commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/abstraction/dto/ComprehendRequest.java:
```diff
@@ -0,0 +1,56 @@
+
+        public String getText() { return text; }
+        public String getLanguageCode() { return languageCode; }
+    }
```
All these classes could be record.

abhishek-gupta-nuxeo replied:
Yes, got that. Will try converting all the newly introduced dto's to a record.

kevinleturc commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/abstraction/AWSServiceRegistry.java:
```diff
@@ -0,0 +1,123 @@
+ * Service implementations only need to inject this registry and access facades through it.
+ * All AWS SDK dependencies are completely isolated in facade implementations.
+ */
+public class AWSServiceRegistry extends DefaultComponent {
+
+    private ComprehendServiceFacade comprehendFacade;
+    private RekognitionServiceFacade rekognitionFacade;
+    private TextractServiceFacade textractFacade;
+    // private TranscribeServiceFacade transcribeFacade; // Temporarily commented out
+    private TranslateServiceFacade translateFacade;
```
All of these components should be registered as such to Nuxeo Runtime, especially here because they are extending DefaultComponent, which makes think that it is a Nuxeo Component but in the end it is not.
That's the responsability of the Nuxeo Runtime to start/stop components, it also brings other features that you don't have to deal with, for instance we can't tell for sure that this registry is handling hot reload properly.

kevinleturc commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/dto/DocumentAnalysisResult.java:
```diff
@@ -0,0 +1,153 @@
+/**
+ * Domain-specific DTO for Textract document analysis results.
+ */
+public class DocumentAnalysisResult {
```
It seems that these classes present in this file are good candidate for records.

Note that record is less verbose and brings out of the box the equals/hashCode/toString methods.

kevinleturc commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/dto/EntitiesResult.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * AWS SDK-independent result DTO for entity detection operations.
+ */
+public class EntitiesResult {
```
Record.

kevinleturc reviewed (commented):
There's a lot of classes (mainly DTO) that could be a record, I made some comment about it but didn't do all the classes.

abhishek-gupta-nuxeo committed:
jdk version upgrade

abhishek-gupta-nuxeo committed:
Optimized code

abhishek-gupta-nuxeo committed:
Merge branch '2025' into task-NXCON-61-release-nuxeo-ai-2025

abhishek-gupta-nuxeo committed:
Getting protobuf-java version from the parent pom

pmacius committed:
OPSEXP-3635 Use reusable nuxeo docker build action (#681)

abhishek-gupta-nuxeo committed:
Nuxeo Base image version update

ataillefer commented on pom.xml:
```diff
@@ -4,27 +4,32 @@
     <groupId>org.nuxeo</groupId>
     <artifactId>nuxeo-parent</artifactId>
-    <version>2023.37.10</version>
+    <version>2025.7</version>
```
Do you want to upgrade to the latest release 2025.11?

abhishek-gupta-nuxeo replied:
Yes, we can proceed with that since it’s the LTS-2025 release. It’s better to use the latest version.

ataillefer commented on .github/workflows/build_and_test.yml:
```diff
@@ -69,8 +69,10 @@ jobs:
+              echo "java-version=17" >> $GITHUB_OUTPUT
           else
-                echo "java-version=17" >> $GITHUB_OUTPUT
+              echo "java-version=21" >> $GITHUB_OUTPUT
```
Since we are in the 2025 branch, do we need these checks? Can't we just stick wih Java 21?

abhishek-gupta-nuxeo replied:
Agreed — in the 2025 branch we can drop the checks and just use Java 21

ataillefer commented on .github/workflows/pipeline_pr.yaml:
```diff
@@ -8,9 +8,9 @@ permissions:
 on:
   pull_request:
-    branches: [master, master-10.10, release-*, '2023']
+    branches: [master, master-10.10, release-*, '2025']
```
Would it make sense to remove all branches except for 2025?

ataillefer commented on .github/workflows/pipeline_pr.yaml:
```diff
@@ -8,9 +8,9 @@ permissions:
+    branches: [master, master-10.10, release-*, '2025']
   push:
-    branches: [master, master-10.10, '2023']
+    branches: [master, master-10.10, '2025']
```
Would it make sense to remove all branches except for 2025?

abhishek-gupta-nuxeo replied:
Yes, since this PR targets the 2025 base branch, we can remove the other branches and keep only 2025.

ataillefer commented on .github/workflows/pipeline_release.yml:
```diff
@@ -20,6 +20,7 @@ on:
           - "8"
           - "11"
           - "17"
+          - "21"
```
Do we need these options? Aka there are cases where we would want to build on a JDK version other than 21?

abhishek-gupta-nuxeo replied:
Good point — we’re only building on JDK 21, so the other options can be removed.

ataillefer reviewed (commented):
OK for th CI part I've reviewed.

abhishek-gupta-nuxeo committed:
Build platform updated

abhishek-gupta-nuxeo committed:
Simplify workflows: remove old branch checks, use only 2025 branch, and standardize on JDK 21.

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/AWSClientFactory.java:
```diff
@@ -0,0 +1,211 @@
+        if (comprehendClient == null) {
+            synchronized (this) {
+                if (comprehendClient == null) {
+                    comprehendClient = ComprehendClient.builder()
+                                                       .region(region)
+                                                       .credentialsProvider(
+                                                               AWSHelper.getInstance().getCredentialsProvider())
+                                                       .build();
```
This pattern is intentionally using double-checked locking with a volatile reference, which is safe. The object is fully constructed before being published, so there is no race condition here. Copilot’s warning is a false positive.

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/AWSClientFactory.java:
```diff
@@ -0,0 +1,211 @@
+        if (rekognitionClient == null) {
+            synchronized (this) {
+                if (rekognitionClient == null) {
+                    rekognitionClient = RekognitionClient.builder()
+                                                         .region(region)
+                                                         .credentialsProvider(
+                                                                 AWSHelper.getInstance().getCredentialsProvider())
+                                                         .build();
+                    log.debug("Created new RekognitionClient");
```
This is safe—volatile plus the synchronized block prevents any visibility or reordering issues, and the AWS client is fully built before assignment. Copilot’s warning is a false positive, so keeping the current code.

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/AWSClientFactory.java:
```diff
@@ -0,0 +1,211 @@
+        if (textractClient == null) {
+            synchronized (this) {
+                if (textractClient == null) {
+                    textractClient = TextractClient.builder()
+                                                   .region(region)
+                                                   .credentialsProvider(
+                                                           AWSHelper.getInstance().getCredentialsProvider())
+                                                   .build();
+                    log.debug("Created new TextractClient");
```
This is the same false positive as above — the field is volatile, initialization happens inside the synchronized block, and the client is fully built before assignment

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/AWSClientFactory.java:
```diff
@@ -0,0 +1,211 @@
+        if (translateClient == null) {
+            synchronized (this) {
+                if (translateClient == null) {
+                    translateClient = TranslateClient.builder()
+                                                     .region(region)
+                                                     .credentialsProvider(
+                                                             AWSHelper.getInstance().getCredentialsProvider())
+                                                     .build();
```
Same as the earlier comments — the combination of volatile, the synchronized block, and the fully constructed AWS client returned by .build() makes this pattern safe.

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/AWSClientFactory.java:
```diff
@@ -0,0 +1,211 @@
+        if (transcribeClient == null) {
+            synchronized (this) {
+                if (transcribeClient == null) {
+                    transcribeClient = TranscribeClient.builder()
+                                                       .region(region)
+                                                       .credentialsProvider(
+                                                               AWSHelper.getInstance().getCredentialsProvider())
+                                                       .build();
```
Same as the earlier comments — the combination of volatile, the synchronized block, and the fully constructed AWS client returned by .build() makes this pattern safe.

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/aws/AWSClientFactory.java:
```diff
@@ -0,0 +1,211 @@
+        if (snsClient == null) {
+            synchronized (this) {
+                if (snsClient == null) {
+                    snsClient = SnsClient.builder()
+                                         .region(region)
+                                         .credentialsProvider(AWSHelper.getInstance().getCredentialsProvider())
+                                         .build();
```
This class uses the correct double-checked locking pattern with volatile. AWS clients are immutable after construction, so no partially initialized state can be observed.

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-gcp-core/src/main/java/org/nuxeo/ai/gcp/GCPClientFactory.java:
```diff
@@ -0,0 +1,156 @@
+                                                                                        FixedCredentialsProvider.create(
+                                                                                                credentials))
+                                                                                .build();
+                        imageAnnotatorClient = ImageAnnotatorClient.create(settings);
```
This warning is a false positive.
The field is volatile and initialized with proper double-checked locking, and the GCP client is fully constructed before assignment.

abhishek-gupta-nuxeo commented on nuxeo-ai-pipes/src/main/java/org/nuxeo/ai/pipes/services/JacksonUtil.java:
```diff
@@ -152,6 +301,70 @@ public void serialize(ManagedBlob blob, JsonGenerator jg, SerializerProvider ser
+                            } catch (Exception ignore) {
+                            }
+                        }
+                        return proxy == args[0];
```
args can’t be null here — the earlier null check guarantees it.

abhishek-gupta-nuxeo commented on addons/nuxeo-ai-aws-core/src/main/java/org/nuxeo/ai/enrichment/DetectTextEnrichmentProvider.java:
```diff
@@ -18,76 +18,51 @@
+    public static final String DEFAULT_CONFIDENCE = "80";
 
-    protected Set<String> textTypes;
+    protected float minConfidence = Float.parseFloat(DEFAULT_CONFIDENCE);
```
This is safe — DEFAULT_CONFIDENCE is a fixed constant ("80"), so Float.parseFloat cannot throw here.

abhishek-gupta-nuxeo committed:
Merge branch '2025' into task-NXCON-61-release-nuxeo-ai-2025

abhishek-gupta-nuxeo committed:
Handled NumberFormatException

abhishek-gupta-nuxeo committed:
Project Version set to 5.0.0-SNAPSHOT

abhishek-gupta-nuxeo committed:
Fix blob handling in Rekognition service to prevent null image data Use AWSHelper.getImage() instead of simplified blob conversion methods to properly handle both S3 and direct blob storage. This fixes the 'Either image data or S3 reference must be provided' error when uploading files.

abhishek-gupta-nuxeo committed:
Sonarcloud code quality issues fixed

abhishek-gupta-nuxeo committed:
nuxeo-ai-model bundle added as required in nuxeo-ai-similar-content manifest file

abhishek-gupta-nuxeo committed:
NXCON-165: Fix 400 error from Sightengine API in nuxeo-ai-image-quality module during image quality check.

- Fix 400 error from Sightengine API during image quality check in nuxeo-ai-image-quality module
- Update nuxeo-jwt-authenticator dependency to version 2.0.0

abhishek-gupta-nuxeo committed:
trigger CI.
